File: person-service/src/main/java/com/microservice/person/model/Address.java
package com.microservice.person.model;

import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import jakarta.validation.constraints.NotEmpty;
import java.util.UUID;

@Data
@RequiredArgsConstructor
@NoArgsConstructor
@Document(collection = "address")
public class Address {
    @Id
    private String id = UUID.randomUUID().toString();
    @NonNull @NotEmpty
    private String address;
    @NonNull @NotEmpty
    private String city;
    @NonNull @NotEmpty
    private String stateOrProvince;
    @NonNull @NotEmpty
    private String country;
    @NonNull @NotEmpty
    private String postalCode;
}

File: scala-address-service/src/main/scala/com/microservice/address/models/Address.scala
package com.microservice.address.models

import java.time.Instant
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.mapping.Document
import io.circe.generic.auto._
import io.circe.{Encoder, Decoder}
import scala.annotation.meta.field

@Document(collection = "addresses")
case class Address(
  @(Id @field) id: Option[String] = None,
  street: String,
  city: String,
  state: String,
  zipCode: String,
  country: String,
  createdByUser: Option[String] = None,
  createdDate: Option[Instant] = Some(Instant.now()),
  lastModifiedByUser: Option[String] = None,
  lastModifiedDate: Option[Instant] = Some(Instant.now())
)

object AddressJsonProtocol {
  implicit val instantEncoder: Encoder[Instant] = Encoder.encodeString.contramap(_.toString)
  implicit val instantDecoder: Decoder[Instant] = Decoder.decodeString.map(Instant.parse)
}
